### DISCUSSION TITLE
Create a full readable documentation

### DISCUSSION THREAD

**Discussion by o-murphy** [o-murphy/py-ballisticcalc#127] [Announcements]
I think we can use somethink like [readthedocs](https://docs.readthedocs.io/en/stable/tutorial/) or [pdoc3](https://github.com/pdoc3/pdoc)

**Comment by o-murphy** [comment]
we would use [mkdocs](https://www.mkdocs.org/)
the template already in repo

**Comment by dbookstaber** [comment]
I was looking into how we can get `mkdocs` + `mkdocstrings` to automatically cross-link objects.

## Current Support

Presently we include the `autorefs` plugin in `mkdocs.yml`.  With this we can use reference-style links in docstrings or Markdown *for objects that have short-name inventory entries*.  Which objects are those?  Hard to predict.  Seems to include classes that are included in the docs, but not members.  So:
- `[Shot][]` << works
- `[Shot.cant_angle][]` << doesn't work
- `[py_ballisticcalc.shot.Shot.cant_angle][]` << works

### Example
Consider the following docstring for the `Shot.cant_angle` attribute:

```python
    cant_angle: Angular
    """Tilt of gun from vertical. If `weapon.sight_height != 0` then this shifts any barrel elevation
            from the vertical plane into the horizontal plane (as `barrel_azimuth`) by `sine(cant_angle)`."""
```

To get cross-links here we would have to write:

```python
    cant_angle: Angular
    """Tilt of gun from vertical. If [`py_ballisticcalc.munition.Weapon.sight_height`][]` != 0` then this shifts any barrel elevation
            from the vertical plane into the horizontal plane (as [`py_ballisticcalc.shot.Shot.barrel_azimuth`][]) by `sine(cant_angle)`."""
```

Crazy, right?

For alternatives I had to go to chatbots since the documentation on these documentation tools is not great.

---

## 1. Central link reference definitions file
Keep short forms like `[Weapon.sight_height][]` in docstrings, but define link targets once.

1. Create (or reuse) a shared reference file, e.g. `docs/_links.md:`

```
[Weapon.sight_height]: ../api/munition/weapon/#py_ballisticcalc.munition.Weapon.sight_height
[Shot.barrel_azimuth]: ../api/conditions/shot/#py_ballisticcalc.shot.Shot.barrel_azimuth
```

(Adjust relative paths if this file ends up nested differently.)

2. Include that file somewhere after all `mkdocstrings` content on every page that needs the links. Easiest: add one global include at the end of each API wrapper markdown file, e.g. in `shot.md` we would include the `pymdownx.snippet`:

```
--8<-- "_links.md"
```

Or maybe that can be included on a global template?

---

## 2. Override template

By default the anchor exposed by `mkdocstrings` for an attribute is usually the full path (`py_ballisticcalc.shot.Shot.cant_angle`). To make short references like `[cant_angle][]` work, create duplicate anchors. Two ways:

### 1a. Override template (official, clean)

1. Create a folder for custom templates, e.g. `docs/_templates/`.
2. Copy the Python handler’s Material template (usually from the `mkdocstrings` repo: `.../handlers/python/templates/material/attribute.html`) into your `_templates`.
4. Add extra `<a id="cant_angle"></a>` (and maybe `<a id="Shot.cant_angle"></a>`) next to the existing anchor:

```jinja2
{# existing anchor (simplified) #}
<a id="{{ identifier }}"></a>
{# add short aliases #}
{% if identifier.endswith('.' + obj.name) %}
  <a id="{{ obj.name }}"></a>
  <a id="{{ obj.parent.name }}.{{ obj.name }}"></a>
{% endif %}
```

5. Point `mkdocstrings` at your template directory:

```yaml
plugins:
  - mkdocstrings:
      handlers:
        python:
          options:
            docstring_style: google
          custom_templates: docs/_templates
```

Now `[cant_angle][]` and `[Shot.cant_angle][]` both resolve.

Pros: No docstring clutter beyond the `[name][]` pattern.  
Cons: One-time template maintenance.

---

## 3. Pre-build rewrite plugin (automatic conversion of backticks)

If you really want to type only `cant_angle` in docstrings and have it link automatically, write a tiny MkDocs plugin to transform occurrences into `[cant_angle][]` (only when they match known symbols).

Skeleton plugin (`docs/plugins/autolink_attrs.py`):

```python
import re
from mkdocs.plugins import BasePlugin

CODE_REF_RE = re.compile(r"(?<!\[)`([A-Za-z_][A-Za-z0-9_\.]*)`(?!\])")

# Provide a whitelist to avoid false positives (populate dynamically if you like)
KNOWN = {
    "Shot", "ShotProps",
    "Shot.cant_angle", "Shot.look_angle", "Shot.relative_angle",
    "cant_angle", "look_angle", "relative_angle",
    # add more...
}

class AutoLinkAttrsPlugin(BasePlugin):
    def on_page_markdown(self, markdown, page, config, files):
        def repl(match):
            name = match.group(1)
            if name in KNOWN:
                return f"[`{name}`][]"
            return match.group(0)
        return CODE_REF_RE.sub(repl, markdown)
```

Enable it:

```yaml
plugins:
  - search
  - mkdocstrings:
      handlers:
        python:
          options:
            docstring_style: google
            merge_init_into_class: true
  - autorefs
  - autolink-attrs
```

Pros: You keep writing plain backticks.  
Cons: Need to maintain `KNOWN` or implement dynamic symbol discovery (can introspect py_ballisticcalc with `inspect` during config load, but that adds import-time side effects if your package does heavy work).

---

## 4. Automated alias hook

Add hooks module that injects alias anchors.  To `mkdocs.yml`:

```
hooks:
  - docs/hooks.py
```

However, this depends on the generated anchors which may differ per template.

**Comment by dbookstaber** [comment]
More `mkdocs` kludginess: `mkdocstrings` (Python handler via Griffe) lists only members that Griffe sees as documented objects: methods, properties with docstrings, etc.  Plain dataclass fields without their own docstrings aren’t emitted as separate items in the TOC, even if described in the class docstring’s Attributes section.  So take a look at [the API page for the Shot class](https://portfolio.o-murphy.net/py-ballisticcalc/latest/api/conditions/shot/#py_ballisticcalc.conditions.Shot) and note:

1. Four attributes are listed in the TOC and in the page body, but only because they are explicit `@property`s with their own getter docstrings.  Other important attributes like `cant_angle` are not listed in the TOC.  Further, only the attributes listed in the TOC can be cross-linked!
2. We can get the other dataclass attributes to appear the same way if we put a docstring *after* the attribute where it is declared in the class body, as shown in the previous comment with the `cant_angle` example.  However, this is does not appear to comply with PEP or docstring standards.
3. Did you notice that any docstring for `@property .setter` method is ignored?  We have to put any `.setter` notes in the `getter` docstring!
